cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTable.java

```diff
@@ -0,0 +1,469 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.storage.spanner;
+
+import com.google.cloud.ByteArray;
+import com.google.cloud.spanner.Key;
+import com.google.cloud.spanner.KeySet;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Statement;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.TransactionContext;
+import com.google.cloud.spanner.Value;
+import io.cdap.cdap.api.dataset.lib.AbstractCloseableIterator;
+import io.cdap.cdap.api.dataset.lib.CloseableIterator;
+import io.cdap.cdap.spi.data.InvalidFieldException;
+import io.cdap.cdap.spi.data.StructuredRow;
+import io.cdap.cdap.spi.data.StructuredTable;
+import io.cdap.cdap.spi.data.table.StructuredTableSchema;
+import io.cdap.cdap.spi.data.table.field.Field;
+import io.cdap.cdap.spi.data.table.field.FieldType;
+import io.cdap.cdap.spi.data.table.field.FieldValidator;
+import io.cdap.cdap.spi.data.table.field.Fields;
+import io.cdap.cdap.spi.data.table.field.Range;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link StructuredTable} implementation backed by GCP Cloud Spanner.
+ */
+public class SpannerStructuredTable implements StructuredTable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SpannerStructuredTable.class);
+
+  private final TransactionContext transactionContext;
+  private final StructuredTableSchema schema;
+  private final FieldValidator fieldValidator;
+
+  public SpannerStructuredTable(TransactionContext transactionContext, StructuredTableSchema schema) {
+    this.transactionContext = transactionContext;
+    this.schema = schema;
+    this.fieldValidator = new FieldValidator(schema);
+  }
+
+  @Override
+  public void upsert(Collection<Field<?>> fields) throws InvalidFieldException {
+    Map<String, Field<?>> fieldMap = fields.stream().collect(Collectors.toMap(Field::getName, Function.identity()));
+    List<Field<?>> primaryKeyFields = new ArrayList<>();
+
+    for (String key : schema.getPrimaryKeys()) {
+      Field<?> field = fieldMap.get(key);
+      if (field == null) {
+        throw new InvalidFieldException(schema.getTableId(), key, "Missing primary key field " + key);
+      }
+      primaryKeyFields.add(field);
+    }
+
+    // Cloud Spanner doesn't support upsert. The best we can do is to read the existing row and update it if it exists
```

Comment:
upsert not supported via DML but available via mutation (i.e. newInsertOrUpdateBuilder) 
https://cloud.google.com/spanner/docs/getting-started/java#write-data-with-mutations

Reply:
We can't use mutation in CDAP due to it cannot read its own writes.